### PR TITLE
Tests for `circuit.py`

### PR DESCRIPTION
**Pull request by isolatedinformation**
Tests to write
- [x] circuit equality
- [x] circuit inequality 
- [x] circuit_has_measurement
- [x] apply_transformation::Commutations between PauliRotations
- [x] are_commuting
- [x] join
- [x] count_rotations_by

### CODE REVIEW

**Comment by alexnguyenn** [comment]
it is better to do 
```
git checkout dev
git pull
git checkout your-branch
git rebase dev
git push -f 
```
Which will give you a cleaner history on the PR (might want to avoid this if there is another person working on the branch).

**Comment by isolatedinformation** [comment]
Tests Remaining:

- [ ]  apply_transformation
     - [ ]  Commutations between PauliRotation and Measurement
- [ ] remove_y_operators_from_circuit
- [ ] commute_pi_over_four_rotations

These will be solved as a part of #156

File: tests/pauli_rotations/generate_tests_circuit.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    `PauliOpCircuit.commute_pi_over_four_rotations()` are also implicitly tested through these tests
+    """
+    tests_list = list()
+    # Test 1:
+    input = PauliOpCircuit(4)
+    input.add_pauli_block(PauliRotation.from_list([X, I, Z, I], Fraction(1, 4)))
+    input.add_pauli_block(PauliRotation.from_list([I, Z, I, X], Fraction(-1, 8)))
+    input.add_pauli_block(PauliRotation.from_list([X, X, I, I], Fraction(-1, 4)))
+    input.add_pauli_block(PauliRotation.from_list([I, I, Z, X], Fraction(1, 8)))
+    input.apply_transformation()
```
Comment (alexnguyenn) [inline tests/pauli_rotations/generate_tests_circuit.py]: I don't think it is a good idea to call the function that is being tested inside here. I suggest passing that `input` circuit as an input to `pytest.parametrize`, and inside the test call:

```
assert input.apply_transformation == expected
```

File: tests/pauli_rotations/generate_tests_circuit.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    c2.add_pauli_block(PauliRotation.from_list([I, I, Z, X], Fraction(1, 8)))
+    c2.add_pauli_block(PauliRotation.from_list([X, X, I, I], Fraction(-1, 4)))
+    c2.add_pauli_block(PauliRotation.from_list([X, I, Z, I], Fraction(1, 4)))
+
+    c3 = PauliOpCircuit(5)
+    c3.add_pauli_block(PauliRotation.from_list([Z, X, Z, Y, I], Fraction(-1, 8)))
+    c3.add_pauli_block(PauliRotation.from_list([Y, Z, Z, X, I], Fraction(1, 8)))
+    c3.add_pauli_block(PauliRotation.from_list([Y, Z, I, Z, Z], Fraction(1, 4)))
+
+    j1 = PauliOpCircuit.join(c1, c2)
```
Comment (alexnguyenn) [inline tests/pauli_rotations/generate_tests_circuit.py]: Same thing with testing `apply_transformation`

File: tests/pauli_rotations/generate_tests_circuit.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    c1.add_pauli_block(PauliRotation.from_list([X, I, Z, I], Fraction(1, 4)))
+    c1.add_pauli_block(PauliRotation.from_list([I, Z, I, X], Fraction(-1, 8)))
+    c1.add_pauli_block(PauliRotation.from_list([X, X, I, I], Fraction(-1, 4)))
+    c1.add_pauli_block(PauliRotation.from_list([I, I, Z, X], Fraction(1, 8)))
+    c1.add_pauli_block(Measurement.from_list([Z, Z, I, I]))
+
+    tests_list.append((c1.count_rotations_by(Fraction(1, 4)), 1))
+    tests_list.append((c1.count_rotations_by(Fraction(-1, 4)), 1))
+    tests_list.append((c1.count_rotations_by(Fraction(1, 8)), 1))
+    tests_list.append((c1.count_rotations_by(Fraction(1, -8)), 1))
```
Comment (alexnguyenn) [inline tests/pauli_rotations/generate_tests_circuit.py]: Same thing with testing `apply_transformation`

File: src/lsqecc/pauli_rotations/circuit.py [hunk 4]
```diff
@@ -356,7 +356,12 @@ def load_reversible_from_qasm_string(quasm_string: str) -> "PauliOpCircuit":
 
     @staticmethod
     def join(lhs: "PauliOpCircuit", rhs: "PauliOpCircuit") -> "PauliOpCircuit":
-        assert lhs.qubit_num == rhs.qubit_num
+        try:
+            assert lhs.qubit_num == rhs.qubit_num
+        except AssertionError:
+            print("The two circuits joined must have the same number of qubits!")
+            # Dummy return for making tests
+            return lhs
```
Comment (alexnguyenn) [inline src/lsqecc/pauli_rotations/circuit.py]: I think we should not omit the error here. You can have a test case that trigger the error and use `with pytest.raises(your_error_type):`

**Review by alexnguyenn — COMMENTED** [review]
Great work for the test cases overall. I have left some comments and suggestions.

File: tests/pauli_rotations/generate_tests_circuit.py (hunk 1 above)
Comment (alexnguyenn) [reply tests/pauli_rotations/generate_tests_circuit.py]: sorry - `apply_transformation()` does not return anything so it should be something like this:
```
input.apply_transformation()
assert input == expected
```

File: tests/pauli_rotations/generate_tests_circuit.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    3. Test 3 --> P, P' commute - Litinski 4b
+    4. Test 4 --> P,P' anti commute - Litinski 4b
+    5. Test 5 --> Controlled Operations, commute - Litinski 4c
+        Tests for this is implicitly included in tests 1 and 2, as controlled operations
+        are converted Pauli Rotations in load_*() methods
+    6. Test 6 --> Controlled Operations, anticommute - Litinski 4c
+        Tests for this is implicitly included in tests 1 and 2, as controlled operations
+        are converted Pauli Rotations in load_*() methods
+
+    `PauliOpCircuit.commute_pi_over_four_rotations()` are also implicitly tested through these tests
```
Comment (alexnguyenn) [inline tests/pauli_rotations/generate_tests_circuit.py:88]: Per https://github.com/latticesurgery-com/lattice-surgery-compiler/issues/156#issuecomment-973179680, I suggest that we add tests for `PauliOpCircuit.commute_pi_over_four_rotations()`, given that it is an important building block for other algorithms and is used more widely than `apply_transformation()`. Though, it should be another PR.

File: src/lsqecc/pauli_rotations/circuit.py (hunk 4 above)
Comment (alexnguyenn) [reply src/lsqecc/pauli_rotations/circuit.py]: I am thinking:
```suggestion
        if lhs.qubit_num != rhs.qubit_num:
            raise Exception("The two circuits joined must have the same number of qubits!")
```
And the test case would be:
```
def test_join_different_qubit_num:
    # some code to initialize the test
    with pytest.raises(Exception):
        # Calling join() on conditions that will raise the exception 
```

File: tests/pauli_rotations/generate_tests_circuit.py (hunk 5 above)
Comment (isolatedinformation) [reply tests/pauli_rotations/generate_tests_circuit.py:88]: Agreed. Let's merge this branch with the changes(after review) and then solve #156 on a seperate PR

File: tests/pauli_rotations/generate_tests_circuit.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    c3.add_pauli_block(PauliRotation.from_list([X, Y, Z, I], Fraction(-1, 4)))
+    c3.add_pauli_block(PauliRotation.from_list([I, Z, Z, I], Fraction(1, 8)))
+    c3.add_pauli_block(Measurement.from_list([Y, X, I, I]))
+    if case == "eq":
+        return [(c1, c1), (c2, c2), (c3, c3)]
+    elif case == "ne":
+        return [(c1, c2), (c2, c3), (c3, c1)]
+    else:
+        assert False
+
```
Comment (alexnguyenn) [inline tests/pauli_rotations/generate_tests_circuit.py]:
```suggestion
def generate_tests_pauli_op_circuit_equality(
    for_eq: bool,
) -> List[Tuple[PauliOpCircuit, PauliOpCircuit]]:
    c1 = PauliOpCircuit(4)
    c1.add_pauli_block(PauliRotation.from_list([I, X, I, I], Fraction(1, 8)))
    c2 = PauliOpCircuit(4)
    c2.add_pauli_block(PauliRotation.from_list([X, Y, Z, I], Fraction(-1, 4)))
    c2.add_pauli_block(PauliRotation.from_list([I, Z, Z, I], Fraction(1, 8)))
    c3 = PauliOpCircuit(4)
    c3.add_pauli_block(PauliRotation.from_list([X, Y, Z, I], Fraction(-1, 4)))
    c3.add_pauli_block(PauliRotation.from_list([I, Z, Z, I], Fraction(1, 8)))
    c3.add_pauli_block(Measurement.from_list([Y, X, I, I]))
    if for_eq:
        return [(c1, c1), (c2, c2), (c3, c3)]
    else:
        return [(c1, c2), (c2, c3), (c3, c1)]
```

File: tests/pauli_rotations/circuit_test.py [hunk 7]
```diff
@@ -1,44 +1,62 @@
...
+    generate_tests_count_rotations_by,
+    generate_tests_join_different_qubit_num,
+    generate_tests_join_same_qubit_num,
+    generate_tests_pauli_op_circuit_equality,
+)
 
 
 @pytest.mark.parametrize(
-    "pauli_op_circuit_1, pauli_op_circuit_2", generate_test_pauli_op_circuit("eq")
+    "pauli_op_circuit_1, pauli_op_circuit_2", generate_tests_pauli_op_circuit_equality("eq")
```
Comment (alexnguyenn) [inline tests/pauli_rotations/circuit_test.py]:
```suggestion
    "pauli_op_circuit_1, pauli_op_circuit_2", generate_tests_pauli_op_circuit_equality(for_eq=True)
```

File: tests/pauli_rotations/circuit_test.py [hunk 8]
```diff
@@ -1,44 +1,62 @@
...
 )
 def test_pauli_op_circuit_eq(pauli_op_circuit_1, pauli_op_circuit_2):
     assert pauli_op_circuit_1 == pauli_op_circuit_1
 
 
 @pytest.mark.parametrize(
-    "pauli_op_circuit_1, pauli_op_circuit_2", generate_test_pauli_op_circuit("ne")
+    "pauli_op_circuit_1, pauli_op_circuit_2", generate_tests_pauli_op_circuit_equality("ne")
```
Comment (alexnguyenn) [inline tests/pauli_rotations/circuit_test.py]:
```suggestion
    "pauli_op_circuit_1, pauli_op_circuit_2", generate_tests_pauli_op_circuit_equality(for_eq=False)
```

File: tests/pauli_rotations/generate_tests_circuit.py (hunk 6 above)
Comment (alexnguyenn) [reply tests/pauli_rotations/generate_tests_circuit.py]: Minor edits to get rid of the `assert False`

**Review by alexnguyenn — APPROVED** [review]
Just some a minor stuff below (you can use `Apply suggestion to batch` and then commit them).

File: tests/pauli_rotations/circuit_test.py [hunk 9]
```diff
@@ -1,44 +1,62 @@
...
+    generate_tests_pauli_op_circuit_equality,
+)
 
 
 @pytest.mark.parametrize(
-    "pauli_op_circuit_1, pauli_op_circuit_2", generate_test_pauli_op_circuit("eq")
+    "pauli_op_circuit_1, pauli_op_circuit_2", generate_tests_pauli_op_circuit_equality("eq")
 )
 def test_pauli_op_circuit_eq(pauli_op_circuit_1, pauli_op_circuit_2):
     assert pauli_op_circuit_1 == pauli_op_circuit_1
```
Comment (alexnguyenn) [inline tests/pauli_rotations/circuit_test.py]:
```suggestion
    assert pauli_op_circuit_1 == pauli_op_circuit_2
```
Also missed this
